core/state-machine/src/testing.rs

```diff
@@ -91,81 +117,109 @@ impl<H: Hasher> Default for TestExternalities<H> {
-// things are defined seems utterly odd to (put changes in changes but never make them
-// available for read through inner)
 impl<H: Hasher> Externalities<H> for TestExternalities<H> where H::Out: Ord {
 	fn storage(&self, key: &[u8]) -> Option<Vec<u8>> {
-		match key {
-			CODE => self.code.clone(),
-			_ => self.inner.get(key).cloned(),
-		}
+		self.overlay.storage(key).map(|x| x.map(|x| x.to_vec())).unwrap_or_else(||
+			self.backend.storage(key).expect(EXT_NOT_ALLOWED_TO_FAIL))
 	}
 
 	fn original_storage(&self, key: &[u8]) -> Option<Vec<u8>> {
-		self.storage(key)
+		self.backend.storage(key).expect(EXT_NOT_ALLOWED_TO_FAIL)
 	}
 
 	fn child_storage(&self, storage_key: ChildStorageKey<H>, key: &[u8]) -> Option<Vec<u8>> {
-		self.changes.child_storage(storage_key.as_ref(), key)?.map(Vec::from)
+		self.overlay
+			.child_storage(storage_key.as_ref(), key)
+			.map(|x| x.map(|x| x.to_vec()))
+			.unwrap_or_else(|| self.backend
+				.child_storage(storage_key.as_ref(), key)
+				.expect(EXT_NOT_ALLOWED_TO_FAIL)
+			)
 	}
 
 	fn place_storage(&mut self, key: Vec<u8>, maybe_value: Option<Vec<u8>>) {
-		self.changes.set_storage(key.clone(), maybe_value.clone());
-		match key.as_ref() {
-			CODE => self.code = maybe_value,
-			_ => {
-				match maybe_value {
-					Some(value) => { self.inner.insert(key, value); }
-					None => { self.inner.remove(&key); }
-				}
-			}
+		if is_child_storage_key(&key) {
+			panic!("Refuse to directly set child storage key");
 		}
+
+		self.overlay.set_storage(key, maybe_value);
 	}
 
-	fn place_child_storage(&mut self, storage_key: ChildStorageKey<H>, key: Vec<u8>, value: Option<Vec<u8>>) {
-		self.changes.set_child_storage(storage_key.into_owned(), key, value);
+	fn place_child_storage(
+		&mut self,
+		storage_key: ChildStorageKey<H>,
+		key: Vec<u8>,
+		value: Option<Vec<u8>>
+	) {
+		self.overlay.set_child_storage(storage_key.into_owned(), key, value);
 	}
 
 	fn kill_child_storage(&mut self, storage_key: ChildStorageKey<H>) {
-		self.changes.clear_child_storage(storage_key.as_ref());
+		let backend = &self.backend;
+		let overlay = &mut self.overlay;
+
+		overlay.clear_child_storage(storage_key.as_ref());
+		backend.for_keys_in_child_storage(storage_key.as_ref(), |key| {
+			overlay.set_child_storage(storage_key.as_ref().to_vec(), key.to_vec(), None);
+		});
 	}
 
 	fn clear_prefix(&mut self, prefix: &[u8]) {
-		self.changes.clear_prefix(prefix);
-		self.inner.retain(|key, _| !key.starts_with(prefix));
+		if is_child_storage_key(prefix) {
+			panic!("Refuse to directly clear prefix that is part of child storage key");
+		}
+
+		self.overlay.clear_prefix(prefix);
+
+		let backend = &self.backend;
+		let overlay = &mut self.overlay;
+		backend.for_keys_with_prefix(prefix, |key| {
+			overlay.set_storage(key.to_vec(), None);
+		});
 	}
 
 	fn chain_id(&self) -> u64 { 42 }
 
 	fn storage_root(&mut self) -> H::Out {
-		trie_root::<H, _, _, _>(self.inner.clone())
+		// compute and memoize
```

Comment:
probably not related to your current issue, but this function would need an equivalent to core/state-machine/src/ext.rs calls to `child_storage_root_transaction` -> `sync_child_storage_root` that put child change to top for root calculation (I think I may have told you something different earlier).
Did you try using directly Ext code instead of TestExternalities (there start to be less and less differences between both).

Reply:
I tried to but couldn't TestExternalities is structure that contains backend and overlay. Ext does only borrow them.

We could write `type TextExternalities = Ext<H, InMemory, InMemoryStorage...>` and make a function `test_externalitie` that returns a structure which implement get_ext that will return the Ext with borrows.
But that would imply change test that make use of `TestExternalities`

Reply:
I took a look it seems that most test that makes use of TestExternalities they are making use of it with with_externalities. and only with a GenesisConfiguration we should probably be able to change with_externalities to instead of taking an externalities parameter just use GenesisConfigParameter or something like that.